{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "using VCFTool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"../input/738_variants_1000.vcf.gz\""
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_dir = \"../input/\"\n",
    "\n",
    "vcf_738_file_path = joinpath(input_dir, \"738_variants.vcf.gz\")\n",
    "\n",
    "vcf_738_1000_file_path = joinpath(input_dir, \"738_variants_1000.vcf.gz\")\n",
    "\n",
    "vcf_file_path_to_use = vcf_738_1000_file_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search by Variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2-d NDSparse with 1000 values (8 field named tuples):\n",
       "    \u001b[4mDimensions\u001b[24m\u001b[1m#  \u001b[22m\u001b[1mcolname  \u001b[22m\u001b[1mtype\u001b[22m\n",
       "──────────────────\n",
       "1  CHROM    String\n",
       "2  POS      String\n",
       "    \u001b[4mValues\u001b[24m\u001b[1m#   \u001b[22m\u001b[1mcolname  \u001b[22m\u001b[1mtype\u001b[22m\n",
       "───────────────────\n",
       "3   ID       String\n",
       "4   REF      String\n",
       "5   ALT      String\n",
       "6   QUAL     String\n",
       "7   FILTER   String\n",
       "8   INFO     String\n",
       "9   FORMAT   String\n",
       "10  GERM     String"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf = make_vcf_ndsparse(vcf_file_path_to_use)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "T\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"T\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "T\n",
      "A\n",
      "1\n",
      "2\n",
      "Any[\"T\", \"A\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "T\n",
      "G\n",
      "1\n",
      "2\n",
      "Any[\"T\", \"G\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "T\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"T\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "A\n",
      "G\n",
      "1\n",
      "2\n",
      "Any[\"A\", \"G\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"1\", \"2\"]\n",
      "A\n",
      "AACAC,AACACAC\n",
      "1\n",
      "2\n",
      "Any[\"AACAC\", \"AACACAC\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "AC\n",
      "A\n",
      "1\n",
      "2\n",
      "Any[\"AC\", \"A\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"1\", \"1\"]\n",
      "C\n",
      "T\n",
      "1\n",
      "2\n",
      "Any[\"T\", \"T\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "G\n",
      "C\n",
      "1\n",
      "2\n",
      "Any[\"G\", \"C\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "A\n",
      "G\n",
      "1\n",
      "2\n",
      "Any[\"A\", \"G\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "G\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"G\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "T\n",
      "C\n",
      "1\n",
      "2\n",
      "Any[\"T\", \"C\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "CA\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"CA\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "T\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"T\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "CGTGT\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"CGTGT\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "G\n",
      "A\n",
      "1\n",
      "2\n",
      "Any[\"G\", \"A\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "T\n",
      "C\n",
      "1\n",
      "2\n",
      "Any[\"T\", \"C\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "T\n",
      "C\n",
      "1\n",
      "2\n",
      "Any[\"T\", \"C\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "T\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"T\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "A\n",
      "G\n",
      "1\n",
      "2\n",
      "Any[\"A\", \"G\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "T\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"T\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "A\n",
      "G\n",
      "1\n",
      "2\n",
      "Any[\"A\", \"G\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "T\n",
      "C\n",
      "1\n",
      "2\n",
      "Any[\"T\", \"C\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"1\", \"1\"]\n",
      "T\n",
      "C,A\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"C\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0\", \"1\"]\n",
      "C\n",
      "T\n",
      "1\n",
      "2\n",
      "Any[\"C\", \"T\"]\n",
      "\n",
      "\n",
      "SubString{String}[\"0|1\"]\n",
      "A\n",
      "ATAT\n",
      "1\n",
      "2\n"
     ]
    },
    {
     "ename": "LoadError",
     "evalue": "BoundsError: attempt to access 1-element Array{SubString{String},1} at index [2]",
     "output_type": "error",
     "traceback": [
      "BoundsError: attempt to access 1-element Array{SubString{String},1} at index [2]",
      "",
      "Stacktrace:",
      " [1] getindex(::Array{SubString{String},1}, ::Int64) at ./array.jl:809",
      " [2] top-level scope at ./In[11]:29",
      " [3] include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1091"
     ]
    }
   ],
   "source": [
    "# https://www.snpedia.com/index.php/Rs1042522\n",
    "\n",
    "using Dates\n",
    "\n",
    "start_time = now()\n",
    "\n",
    "for value in vcf[\"chr1\", :]\n",
    "    \n",
    "    genotype = []\n",
    "    \n",
    "    allele_ref = value[:REF]\n",
    "    \n",
    "    allele_alt = value[:ALT]\n",
    "    \n",
    "    numeric_genotype = split(split(value[:GERM], \":\")[1], \"/\")\n",
    "    \n",
    "    println(numeric_genotype)\n",
    "    \n",
    "    println(allele_ref)\n",
    "    \n",
    "    println(allele_alt)\n",
    "\n",
    "    counter = 1\n",
    "    \n",
    "        while counter < 3\n",
    "        \n",
    "            println(counter)\n",
    "        \n",
    "            if \"0\" == numeric_genotype[counter]\n",
    "\n",
    "                push!(genotype, allele_ref)\n",
    "\n",
    "            elseif \"1\" == numeric_genotype[counter]\n",
    "\n",
    "                push!(genotype, split(allele_alt, \",\")[1])\n",
    "\n",
    "            elseif \"2\" == numeric_genotype[counter]\n",
    "\n",
    "                push!(genotype, split(allele_alt, \",\")[2])\n",
    "        \n",
    "            end    \n",
    "    \n",
    "        counter += 1\n",
    "        \n",
    "        end\n",
    "\n",
    "    println(genotype)\n",
    "    \n",
    "    println(\"\\n\")\n",
    "    \n",
    "end\n",
    "  \n",
    "end_time = now()\n",
    "\n",
    "println(\"\\nTook $(canonicalize(Dates.CompoundPeriod(end_time - start_time))).\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Pkg; Pkg.add(\"Dagger\")\n",
    "\n",
    "using Dagger\n",
    "\n",
    "Dagger.save(vcf, \"../input/test_vcf_ndsparse_save.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Dagger\n",
    "\n",
    "dagger_load_vcf = Dagger.load(\"../input/test_vcf_ndsparse_save.vcf\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search by Gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search by Region"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tabix Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_file_path = joinpath(input_dir, \"cardiotoxicity.bed\")\n",
    "\n",
    "tabix_regions_from_file(\n",
    "    bed_file_path,\n",
    "    vcf_738_file_path,\n",
    "    output_dir,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## JuliaDB\n",
    "\n",
    "An unzipped VCF file can be loaded into an indextable or ndsparse array directly using loadtable() or ndsparese() respectively. However, its much faster to query variants if they are in an ndsparse array because the chromosome and position are both part of the index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_ndsparse_738 = make_ndsparse(\"../input/738_variants.vcf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for item in vcf_ndsparse_738\n",
    "    \n",
    "    println(item) \n",
    "\n",
    "    println(item[:REF])\n",
    "\n",
    "    end\n",
    "    \n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.3",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
